File: 34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        targetIdx = self.binarySearch(nums, 0, len(nums) - 1, target)
        
        if targetIdx == -1:
            return [-1, -1]
        
        tempRight, tempLeft = targetIdx, targetIdx

        while tempRight != -1:
            startIdx = tempRight
            tempRight = self.binarySearch(nums, 0, tempRight - 1, target)                                
           
        while tempLeft != -1:
            endIdx = tempLeft
            tempLeft = self.binarySearch(nums, tempLeft + 1, len(nums) - 1, target)
            
        return [startIdx, endIdx]
    
        
    def binarySearch(self, nums: List[int], left, right, target) -> int:        
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        
        return -1
                

File: swift/34_find_first_and_last_position_of_element_in_sorted_array.swift
class Solution {
  func searchRange(_ nums: [Int], _ target: Int) -> [Int] {
    let start: Int = findStart(nums, target)
    let end: Int = findEnd(nums, target)

    return [start, end]
  }

  private func findStart(_ nums: [Int], _ target: Int) -> Int {
    var left: Int = 0
    var right: Int = nums.count - 1

    while left <= right {
      let mid = left + (right - left) / 2

      if nums[mid] == target {
        if mid == 0 || nums[mid - 1] != target {
          return mid
        } else {
          right = mid - 1
        }
      } else if nums[mid] < target {
        left = mid + 1
      } else {
        right = mid - 1
      }
    }

    return -1
  }

  private func findEnd(_ nums: [Int], _ target: Int) -> Int {
    var left: Int = 0
    var right: Int = nums.count - 1

    while left <= right {
      let mid = left + (right - left) / 2

      if nums[mid] == target {
        if mid == nums.count - 1 || nums[mid + 1] != target {
          return mid
        } else {
          left = mid + 1
        }
      } else if nums[mid] < target {
        left = mid + 1
      } else {
        right = mid - 1
      }
    }

    return -1
  }
}
